boot/bootstate20.go

```diff
@@ -158,22 +193,38 @@ func (ks20 *bootState20Kernel) commit() error {
 
 	// If we are about to try an update, and need to add the try-kernel symlink,
 	// we need to do things in this order:
-	// 1. Add try-kernel symlink
-	// 2. Update kernel_status to "try"
+	// 1. Add the kernel snap to the modeenv
+	// 2. Create try-kernel symlink
+	// 3. Update kernel_status to "try"
 	//
-	// This is because if we get rebooted in between 1 and 2, kernel_status
-	// is still unset and boot scripts proceeds to boot with the old kernel,
-	// effectively ignoring the try-kernel symlink.
+	// This is because if we get rebooted in before 3, kernel_status is still
+	// unset and boot scripts proceeds to boot with the old kernel, effectively
+	// ignoring the try-kernel symlink.
 	// If we did it in the opposite order however, we would set kernel_status to
 	// "try" and then get rebooted before we could create the try-kernel
 	// symlink, so the bootloader would try to boot from the non-existent
 	// try-kernel symlink and become broken.
+	//
+	// Adding the kernel snap to the modeenv's list of trusted kernel snaps can
+	// effectively happen any time before we update the kernel_status to "try"
+	// for the same reasoning as for creating the try-kernel symlink. Putting it
+	// first is currently a purely aesthetic choice.
 
-	// add the try-kernel symlink
-	// trySnap could be nil here if we called setNext on the current kernel
-	// snap
+	// add the kernel to the modeenv and add the try-kernel symlink
+	// tryKernelSnap could be nil here if we called setNext on the current
+	// kernel snap
 	if ks20.tryKernelSnap != nil {
-		err := ks20.ebl.EnableTryKernel(ks20.tryKernelSnap)
+		// add the kernel to the modeenv
+		ks20.kModeenv.modeenv.CurrentKernels = append(
+			ks20.kModeenv.modeenv.CurrentKernels,
+			ks20.tryKernelSnap.Filename(),
+		)
```

Comment:
So I think we can avoid the attack I mentioned in my comment with minimal changes to this branch if we enforce that current_kernels is ordered, with the fallback first and the try kernel second. So either:
- `current_kernels=pc-kernel_1.snap` where revision 1 is the fallback kernel
- `current_kernels=pc-kernel_1.snap,pc-kernel_2.snap` where revision 1 is the fallback kernel, and revision 2 is the try kernel

I can add tests that this is true in practice, but I'm wondering if we should specifically write that out here instead of appending? I.e. this would become:

```go
		ks20.kModeenv.modeenv.CurrentKernels = []string{
			ks20.currentKernelSnap.Filename(),
			ks20.tryKernelSnap.Filename(),
		}
```

@pedronis WDYT?

Reply:
What is the actual concern here,  by definition we almost always want to be able to boot with the previous kernel given that the new one can fail?  and an attacker can empty boot to brick us or remove any trace of ongoing update, removing try-kernel completely and changing the boot env?

Reply:
Just for clarity to other reviewers as we discussed this elsewhere, the concern is that:
1. modeenv's `current_kernels` currently is not treated as being sorted
1. Since `current_kernels` isn't being treated as sorted, the bootloader and the initramfs don't have a way to identify the situation when the values of kernel.efi and try-kernel.efi are _swapped_
1. The attack of swapping kernel.efi and try-kernel.efi could also essentially be implemented by just attacking the bootloader on ubuntu-boot and always set `kernel_status=trying`, but at least in that attack `snap changes` would report that the kernel was rolled back over reboot and mark the change as `Undone`/`Error`. This symlink swapping is different because the code as I have it written here and in my initramfs branch does not ever detect the problem and so snapd in user space happily thinks it upgraded to the try-kernel.
1. Since the attack can be performed other ways, it's less concerning from a security POV, but I still would rather implement my proposed fix because it at least gives us observability of the modification and blocks boot from happening when the symlinks are swapped.